app/src/main/java/org/schabi/newpipe/database/playlist/dao/PlaylistStreamDAO.java

```diff
@@ -83,4 +83,23 @@ default Flowable<List<PlaylistStreamEntity>> listByService(final int serviceId)
             + " GROUP BY " + JOIN_PLAYLIST_ID
             + " ORDER BY " + PLAYLIST_NAME + " COLLATE NOCASE ASC")
     Flowable<List<PlaylistMetadataEntry>> getPlaylistMetadata();
+
+    @RewriteQueriesToDropUnusedColumns
+    @Transaction
+    @Query("SELECT *, MIN(" + JOIN_INDEX + ") FROM " + STREAM_TABLE + " INNER JOIN "
+            + "(SELECT " + JOIN_STREAM_ID + "," + JOIN_INDEX
+            + " FROM " + PLAYLIST_STREAM_JOIN_TABLE
+            + " WHERE " + JOIN_PLAYLIST_ID + " = :playlistId)"
+            + " ON " + STREAM_ID + " = " + JOIN_STREAM_ID
+            + " LEFT JOIN "
+            + "(SELECT " + JOIN_STREAM_ID + " AS " + JOIN_STREAM_ID_ALIAS + ", "
+            + STREAM_PROGRESS_MILLIS
+            + " FROM " + STREAM_STATE_TABLE + " )"
+            + " ON " + STREAM_ID + " = " + JOIN_STREAM_ID_ALIAS
+            + " GROUP BY " + STREAM_ID
+            + " ORDER BY " + JOIN_INDEX + " ASC")
```

Comment:
Is this technically better?
```suggestion
            + " ORDER BY MIN(" + JOIN_INDEX + ") ASC")
```

Reply:
IMO, the difference in performance is likely to be negligible, especially if the result set is small (which it always is in our case) and the number of groups is large (_duplicates<<streams_). However, if you would like me to implement it, I would be happy to do so.

Reply:
@Stypox Do you have any thoughts on this?

Reply:
I was't talking about difference in performance, but rather about correctness. Since `JOIN_INDEX` is not in the `GROUP BY` clause, what does it mean to order groups by their join index? Take these two groups for example (2,3), (1,4) (the numbers represent the join indices). Which one should come first, when sorting by join index? In theory SQlite should raise an error, I don't know why it doesn't (I think postgresql would). If instead you sort by `MIN(JOIN_INDEX)`, then it's perfectly clear that (1,4) comes before (2,3).

Reply:
Thank you for the example. I understand it and will change it now.